olmo/optim.py

```diff
@@ -144,12 +147,12 @@ def is_grad_norm_metric(metric_name: str) -> bool:
             # Reduce mins.
             if per_param_min_metrics:
                 all_mins = torch.cat(per_param_min_metrics).to(device)
-                dist.reduce(all_mins, 0, op=dist.ReduceOp.MIN)
+                dist.reduce(all_mins, 0, op=dist.ReduceOp.MIN, group=process_group)
```

Comment:
This was the original approach I went with, but I had a few concerns:

1. Rank 0 refers to the global rank, so only the 'first' group will perform the reduce. The code says it will warn for the other groups, and I vaguely remember my runs crashing instead. If it doesn't crash then I guess it's not a real problem. You can get around this by getting each group's rank 0 using `process_group.get_global_rank(group, 0)`.
2. Using multiple process groups on the same stream without some sort of synchronization can lead to deadlocks (https://pytorch.org/docs/stable/distributed.html#torch.distributed.new_group). I don't fully understand it myself. The way torch seems to get around this is that it puts the ops on different streams AND (if dynamo compiling is off) has the streams wait on each other. For us, it may be that we have to either pass the process group to all `dist.*` calls or that we have to synchronize when we make distributed calls over different process groups.

Reply:
@2015aroras thanks, 6911bfb9 should fix point (1). I'll see what happens now about (2).

Reply:
Seems to be working fine on a single node test at least.

Reply:
@2015aroras I'm going to merge and give this a try with the 70B. If I run into other issues I'll rethink this strategy.